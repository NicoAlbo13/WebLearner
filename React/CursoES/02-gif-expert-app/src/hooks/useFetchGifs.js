import { useState, useEffect } from "react";
import getGifs from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    const [data, setData] = useState([]);
    const [isLoading, setIsLoading] = useState(true);

    const getFinalGifs = async ()=>{
        const newGif = await getGifs(category);
        setData(newGif)
        setIsLoading(false)
    }

    useEffect(() => {
        getFinalGifs();
    }, [])


    return {
        images: data,
        isLoading: isLoading
    }
}
