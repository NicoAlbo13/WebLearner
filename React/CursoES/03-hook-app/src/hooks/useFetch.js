import { useEffect, useState } from 'react';

const localCache = {};

export const useFetch = (url) => {

    const [state, setState] = useState({
        data: null,
        isLoading: true,
        hasError: false,
        errorMessage: null,
    })

    useEffect(()=>{
        getFetch()
    },[url])

    const seLoadingState = ()=>{
        setState({
        data: null,
        isLoading: true,
        hasError: false,
        errorMessage: null,
        })  
    }


    const getFetch = async ()=>{
        seLoadingState();

        //if data already on cache retrieve from there
        if(localCache[url]){
            setState({
                data: localCache[url],
                isLoading: false,
                hasError: false,
                errorMessage: null,
            });
            return;
        }

        const response = await fetch(url);

        await new Promise(resolve => setTimeout(resolve, 900))

        if(!response.ok){
            setState({
                data: null,
                isLoading: false,
                hasError: true,
                errorMessage: {
                    code: response.status,
                    message: response.statusText
                },
            })

            return;
        }

        const data = await response.json();
        setState({
            data: data,
            isLoading: false,
            hasError: false,
            errorMessage: null,
        })

        // console.log({data});
        //save to cache
        localCache[url]=data;
    }

  return {
    data: state.data,
    isLoading: state.isLoading,
    hasError: state.hasError
  }
}
