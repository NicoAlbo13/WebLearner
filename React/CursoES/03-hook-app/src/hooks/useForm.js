import { useState } from "react";

export const useForm = (initialForm={}) => {

    const [formState, setFormState] = useState(initialForm)

    
        function handleChange ({target}){
            const {name, value} = target;
            setFormState({
                ...formState, 
                [name]:value
            })
        }

        function onClearForm(e){
            // e.preventDefault()
            setFormState(initialForm)
        }

    return {
        ...formState,
        formState,
        handleChange,
        onClearForm
    }
}
