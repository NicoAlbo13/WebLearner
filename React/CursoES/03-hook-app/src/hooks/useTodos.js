import { useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = ()=>{
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const useTodos = ()=> {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])
    

    const handleNewTodo = (todo)=>{
        const newAction = {
            type: '[TODO] Add Todo',
            payload: todo,
        };
        
        dispatch(newAction);
    }

    const handleDeleteTodo = (id)=>{
        dispatch({
            type: '[TODO] Delete Todo',
            payload: id,
        });
    }

    const handleToggleTodo = (id)=>{        
        dispatch({
            type: '[TODO] Toggle Todo',
            payload: id,
        })
    }
    

    return {
    todos,
    todosCount: todos.length,
    pendingTodoCount: todos.filter(todo=>(!todo.done)).length,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    }
}
