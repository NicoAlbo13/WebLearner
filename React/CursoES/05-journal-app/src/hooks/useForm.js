import { useEffect, useMemo, useState } from 'react';

export const useForm = ( initialForm = {}, formValidator={} ) => {

    const [ formState, setFormState ] = useState( initialForm );
    const [formValidation, setFormValidation] = useState({});

    useEffect(() => {
        createValidators()
    }, [formState])

    useEffect(() => {
        setFormState(initialForm);
    }, [initialForm])

    const isFormValid = useMemo(()=>{
        for (const field of Object.keys(formValidation)) {
            if(formValidation[field] !== null){
                // console.log(formValidation[field]);
                return false;
            }
        }
        return true;
    },[formValidation]) 

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [ name ]: value
        });
    }

    const onResetForm = () => {
        setFormState( initialForm );
    }

    const createValidators = () =>{
        const finalValues = {}

        for (const field of Object.keys(formValidator)) {
            const [fn, message] = formValidator[field];
            finalValues[`${field}Valid`] = fn(formState[field]) ? null : message;
        }

        setFormValidation(finalValues);
    }

    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,

        ...formValidation,
        isFormValid,
    }
}