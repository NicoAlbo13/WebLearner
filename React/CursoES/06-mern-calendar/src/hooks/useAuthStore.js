import { useDispatch, useSelector } from "react-redux"

import calendarApi from "../api/calendarApi";
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store/auth/authSlice";
import { onClearCalendar } from "../store/calendar/calendarSlice";


export const useAuthStore = () => {

    const { status, user, errorMessage } = useSelector(state => state.auth);
    const dispatch = useDispatch();

    const startLogin = async ({email, password}) => {
        dispatch( onChecking() );
        try {
            const res = await calendarApi.post('/auth', { email, password });

            const { token, uid, name } = res.data;
            localStorage.setItem('token', token);
            localStorage.setItem('token-init', new Date().getTime());

            dispatch(onLogin({uid, name}));

        } catch (error) {
            // console.log(error);
            dispatch(onLogout('Incorrect credentials'));
            setTimeout(()=> {
                dispatch(clearErrorMessage());
            }, 10)
        }
    }

    const startRegister = async({name, email, password}) => {
        dispatch(onChecking());

        try {
            const res = await calendarApi.post('/auth/new', {name, email, password});
            const { token, uid } = res.data;
            localStorage.setItem('token', token);
            localStorage.setItem('token-init', new Date().getTime());

            dispatch(onLogin({uid, name}));

        } catch (error) {
            // console.log(error.response.data);
            dispatch(onLogout(error.response.data.msg || 'Error on the register form'));
            setTimeout(()=> {
                dispatch(clearErrorMessage());
            }, 10)
        }
    }

    const checkAuthToken = async() => {
        const token = localStorage.getItem('token');
        if( !token ) return dispatch( onLogout(undefined) );

        try {
            const { data } = await calendarApi.get('/auth/renew');
            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init', new Date().getTime());

            dispatch(onLogin({uid: data.uid, name: data.name}));

        } catch (error) {
            localStorage.clear();
            dispatch( onLogout(undefined) );
        }
    }

    const startLogout = () => {
        localStorage.clear()
        dispatch(onClearCalendar())
        dispatch(onLogout(undefined));
    }

    return {
        //Properties
        status,
        user,
        errorMessage,

        //Methods
        checkAuthToken,
        startLogin,
        startLogout,
        startRegister,
    }
}
