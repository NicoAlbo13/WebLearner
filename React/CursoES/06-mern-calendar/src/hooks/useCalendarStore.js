import { useDispatch, useSelector } from "react-redux"
import { onAddNewEvent, onDeleteEvent, onSetActiveEvent, onUpdateEvent } from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {

    const dispatch = useDispatch();
    const {events, activeEvent} = useSelector(state=>state.calendar);

    const setActiveEvent = ( calendarEvent ) => {
        dispatch(onSetActiveEvent(calendarEvent))
    }

    const startSavingEvent = async(calendarEvent) => {
        //TODO: All backend job etc...
        if(calendarEvent._id){
            //updating note
            dispatch(onUpdateEvent(calendarEvent));
        }else{
            //creating new note
            dispatch(onAddNewEvent({...calendarEvent, _id: new Date().getTime()}))
        }
    }

    const startDeletingEvent = () => {
        dispatch(onDeleteEvent())
    }

    return{
        //Properties
        events,
        activeEvent,
        hasEventSelected: !!activeEvent,

        //Methods
        setActiveEvent,
        startSavingEvent,
        startDeletingEvent,
    }
}
