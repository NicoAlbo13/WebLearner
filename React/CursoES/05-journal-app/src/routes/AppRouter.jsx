import { Navigate, Route, Routes } from 'react-router'
import { AuthRoutes } from '../auth/routes/AuthRoutes'
import { JournalRoutes } from '../journal/routes/JournalRoutes'
import { LoadingAuth } from '../ui'
import { useCheckAuth } from '../hooks'

export const AppRouter = () => {

  const {status} = useCheckAuth()

  if(status === 'checking'){
    return <LoadingAuth/>
  }

  return (
    <Routes>
      {
        status === 'authenticated'
        ?<Route path='/*' element={<JournalRoutes/>}/>
        :<Route path='auth/*' element={<AuthRoutes/>}/>
      }

        <Route path='/*' element={<Navigate to='auth/login'/>}/>
        {/* Login */}
        {/* <Route path='auth/*' element={<AuthRoutes/>}/> */}

        {/* App */}
        {/* <Route path='/*' element={<JournalRoutes/>}/> */}
    </Routes>
  )
}
