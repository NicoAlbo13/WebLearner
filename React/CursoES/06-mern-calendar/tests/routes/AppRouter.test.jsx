import { render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router"
import { AppRouter } from "../../src/routes/AppRouter"
import { useAuthStore } from "../../src/hooks/useAuthStore"

jest.mock("../../src/hooks/useAuthStore")

jest.mock("../../src/calendar/pages/CalendarPage",()=>({
    CalendarPage: ()=> (<h1>CalendarPage</h1>)
}))

describe('test on the <AppRouter/>', () => {

    const mockedCheckAuthToken = jest.fn();

    beforeEach(()=> jest.clearAllMocks())

    test('should show a loading screen and call the checkAuthToken', () => {
        useAuthStore.mockReturnValue({
            checkAuthToken: mockedCheckAuthToken,
            status: 'checking'
        })
        render(<AppRouter/>)
        const title = screen.getByRole('heading', {level: 3})
        expect(title.innerHTML).toBe("Loading...");
        expect(mockedCheckAuthToken).toHaveBeenCalled()
    })

    test('should show the login in case not-authenticated', () => {
        useAuthStore.mockReturnValue({
            checkAuthToken: mockedCheckAuthToken,
            status: 'not-authenticated'
        })

        const {container} =render(
            <MemoryRouter>
                <AppRouter/>
            </MemoryRouter>
        )

        expect(screen.getAllByText('Login')).toBeTruthy();
        expect(container).toMatchSnapshot()
    })

    test('should show the calendar in case authenticated', () => {
        useAuthStore.mockReturnValue({
            checkAuthToken: mockedCheckAuthToken,
            status: 'authenticated'
        })

        const {container} =render(
            <MemoryRouter>
                <AppRouter/>
            </MemoryRouter>
        )
        
        expect(screen.getByText('CalendarPage')).toBeTruthy()
    })

})
