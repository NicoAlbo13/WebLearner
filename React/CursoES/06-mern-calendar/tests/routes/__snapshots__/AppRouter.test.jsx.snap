// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`test on the <AppRouter/> should show the login in case not-authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Login
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create Account"
            />
            <label
              style="color: red;"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
